development/build/transforms/remove-fenced-code.js

```diff
@@ -0,0 +1,425 @@
+    this._fileBuffers = [];
+  }
+
+  // This function is called whenever data is written to the stream.
+  // It concatenates all buffers for the current file into a single buffer.
+  _transform(buffer, _encoding, next) {
+    this._fileBuffers.push(buffer);
+    next();
+  }
+
+  // "flush" is called when all data has been written to the
+  // stream, immediately before the "end" event is emitted.
+  // It applies the transform to the concatenated file contents.
+  _flush(end) {
+    const [fileContent] = removeFencedCode(
+      this.filePath,
+      this.buildType,
+      Buffer.concat(this._fileBuffers).toString('utf8'),
+    );
+
+    this.push(fileContent);
+    end();
+  }
+}
+
+/**
+ * A factory for a Browserify transform that removes fenced code from all
+ * JavaScript source files. The transform is applied to files with the following
+ * extensions:
+ *   - `.js`
+ *   - `.cjs`
+ *   - `.mjs`
+ *
+ * For details on how the transform mutates source files, see
+ * {@link removeFencedCode} and the documentation.
+ *
+ * @param {string} buildType - The type of the current build.
+ * @returns {(filePath: string) => Transform} The transform function.
+ */
+function createRemoveFencedCodeTransform(buildType) {
+  if (!hasOwnProperty(BuildTypes, buildType)) {
+    throw new Error(
+      `Metamask build: Code fencing transform received unrecognized build type "${buildType}".`,
+    );
+  }
+
+  // Browserify transforms are functions that receive a file name and return a
+  // duplex stream. The stream receives the file contents piecemeal in the form
+  // of Buffers.
+  // To apply our code fencing transform, we concatenate all buffers and convert
+  // them to a single string, then apply the actual transform function on that
+  // string.
+  /**
+   * @returns {Transform}
+   */
+  return function removeFencedCodeTransform(filePath) {
+    if (!['.js', '.cjs', '.mjs'].includes(path.extname(filePath))) {
+      return new PassThrough();
+    }
+
+    return new RemoveFencedCodeTransform(filePath, buildType);
+  };
+}
+
+const DirectiveTerminuses = {
+  BEGIN: 'BEGIN',
+  END: 'END',
+};
+
+const DirectiveCommands = {
+  ONLY_INCLUDE_IN: 'ONLY_INCLUDE_IN',
+};
+
+const CommandValidators = {
+  [DirectiveCommands.ONLY_INCLUDE_IN]: (params, filePath) => {
+    if (!params || params.length === 0) {
+      throw new Error(
+        getInvalidParamsMessage(
+          filePath,
+          DirectiveCommands.ONLY_INCLUDE_IN,
+          `No params specified.`,
+        ),
+      );
+    }
+
+    params.forEach((param) => {
+      if (!hasOwnProperty(BuildTypes, param)) {
+        throw new Error(
+          getInvalidParamsMessage(
+            filePath,
+            DirectiveCommands.ONLY_INCLUDE_IN,
+            `"${param}" is not a valid build type.`,
+          ),
+        );
+      }
+    });
+  },
+};
+
+// Matches lines starting with "///:", optionally preceded by whitespace
+const linesWithFenceRegex = /^.*\/\/\/:.*$/gmu;
```

Comment:
This regex doesn't appear to do what is described. First off, it matches on _any_ preceding characters, not just whitespace. But more importantly, this doesn't match lines. The `^` and `$` assertions match the beginning and end of the _input_, not the _line_. And since this is a multi-line regex, this will return multi-line results.

Reply:
I'd recommending using something more like this:

```suggestion
// Matches lines starting with "///:", optionally preceded by whitespace
const lineWithFenceRegex = /^\s*\/\/\/:.*$/u;
```

and doing the line-splitting separately, before using this regex.

Reply:
Great point on the white space. I'm on mobile so I can't check right now, but I can't see how my tests would patch if it matched the beginning and end of the entire input.

Reply:
The `m` flag should make the caret and dollar sign match the beginning and end of lines: https://javascript.info/regexp-multiline-mode

Reply:
Oh, oops, yes you're right. I actually tested this and got the correct results but misread them them. :facepalm:

Reply:
Where is my "😅" react?!

Reply:
```suggestion
// Matches lines starting with "///:", optionally preceded by whitespace
const linesWithFenceRegex = /^\s*\/\/\/:.*$/gmu;
```